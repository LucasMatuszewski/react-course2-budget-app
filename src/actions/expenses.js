import uuid from 'uuid';

// we make here NAMED EXPORTS (not default)

/***********
 * ACTIONS (action generators - functions to generate an Action)
 */

// ADD_EXPENSE ( property = default value )
export const addExpense = (
    {
        description = '',
        note = '',
        amount = 0,
        createdAt = 0
    } = {}
) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuid(), // we use uuid - NPM library to generate RFC4122 UUIDs (we have to install and import it)
        description,
        note,
        amount,
        createdAt
    }
});

// REMOVE_EXPENSE
// const removeExpense = (id) => ({ // return object: {id: "UUID"}
export const removeExpense = ({ id } = {}) => ({ // return only UUID
    type: 'REMOVE_EXPENSE',
    id
    // expense: { id } // we don't need "expense: { }" here. We only want to pass id to delete it
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});