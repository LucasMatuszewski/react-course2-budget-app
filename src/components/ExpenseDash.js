import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';

const ExpenseDash = () => (
    <div>
        <h2>Expenses Dashboard</h2>

        <ExpensesSummary />
        <ExpenseListFilters />
        <ExpenseList /> {/* We don't need to pass any data/props here. connect() will pass them directly from a store / state tree */}
        <br />
        <hr />
        <a href="/help">Server Side Rendering HTML Link</a><br />
        (html "a" tag will reload a browser, we can use it to link to external page)
    </div>
);

export default ExpenseDash;