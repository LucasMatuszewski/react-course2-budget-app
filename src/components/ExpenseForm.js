import React from 'react';
import moment from 'moment'; // http://momentjs.com/ (great library to work with time)
import 'react-dates/initialize'; // As of v13.0.0 this project relies on react-with-styles
import { SingleDatePicker } from 'react-dates'; // AIRbnb date picker: http://airbnb.io/react-dates

// It's better to move _datepicker.css to app.js because we will use it in many components:
// import 'react-dates/lib/css/_datepicker.css'; //obligatory CSS for react-dates

/* 
/////////////////////////////
// JS DATE() vs MOMENT(): //
///////////////////////////
const date = new Date(); //now date: Sun Jun 24 2018 16:46:52 GMT+0200 (Central Europe Daylight Time)
// API of Date() is horrible, thats why we use Moment.js (much better, it's a standard now)
console.log('Date():', date);

const now = moment(); // we have to use it only once.
console.log('moment():', now); // returns object with all moment data and methods
console.log('moment().format():', now.format()); // returns a string with default format: 2018-06-24T16:50:46+02:00
// how to format a date: http://momentjs.com/docs/#/displaying/
console.log(now.format('MMM Do, YYYY')); // US format: Jun 24th, 2018
console.log(now.format('DD-MM-YYYY')); // 24-06-2018
// You can use "-,: " how you prefer.

// TO SET LOCAL SETTINGS OF TIME (GLOBALLY FOR a Component)
//moment.locale('pl'); // IT RETURNS TypeError on JEST test: moment.locale is not a function...
*/


// We use Class StateFull Component because we want to track inner state for data changes inside this form.
// With local state we can validate and keep data in local component state etc.
// We will change Redux Store only on Button submit
export default class ExpenseForm extends React.Component {
    /* // MY VERSION (worked OK without constructor super(props) - but how???)
    state = {
        description: this.props.expense ? this.props.expense.description : '', // only description is obligatory
        note: this.props.expense ? this.props.expense.note : '',
        amount: this.props.expense ? this.props.expense.amount : '',
        createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(), //default: NOW (object from moment.js)
        calendarFocused: false,
        error: undefined
    }; */
    // UDEMY VERSION:
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '', // only description is obligatory
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(), //default: NOW (object from moment.js)
            // moment(MILLISECONDS) = object with calculated date
            calendarFocused: false,
            error: undefined
        };
    }
    onDescriptionChange = (e) => {
        const description = e.target.value; 
        this.setState(() => ({ description }))
    };
/*  THIS VERSION DIDN'T WORK ON JEST TEST WITH ENZYME:   
    onNoteChange = (e) => {
        // 2nd Way to do this (opposite to description):
        // const note = e.target.value; // We can't use it inside call back function for setState, unless we persist it
        e.persist(); // JEST TEST TYPE ERROR: IS NOT A FUNCTION
        this.setState(() => ({ note: e.target.value }));
    }; */
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        // JS FORM VALIDATION (can be double in HTML5 and in JS):
        // if (amount.match(/^\d*(\.\d{0,2})?$/)) { //test for match  with RegEx = Regular Expression.
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) { //empty amount or amount matching RegEx = Regular Expression.
            // always put RegEx inside /.../
            // ^ = beginning of expression, force to start with (in this case) \d = digit
            // \d = digits only (but only one)
            // * = many (0 to infinity), \d* = many digits (but 0 too)
            // {1,} = 1 to infinity \d{1,} = minimum 1 number
            // ()? = optional part of expression
            // \. = allow one dot "." (\.* = many dots)
            // {0,2} = limit from 0 (min) to 2 (max) characters of what is in front of it, \d{0,2} = 0-2 digits, 
            // $ = end of expression
            // Great generator, tester and Library of RegEx: https://regex101.com/
            this.setState(() => ({ amount }));
        } else {
            this.setState(() => ({ error: 'Amount should be in price format: 100.00 or 100' }));
        }
    };
    onAmountChangeX = (e) => {
        const amount = e.target.value;
        this.setState(() => ({ amount }));
    };
    onDateChange = (createdAt) => {
        createdAt && this.setState(() => ({ createdAt })); // createdAt && = prevent from deleting Date
    };
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };
    onSubmit = (e) => {
        e.preventDefault(); // will stop browser refreshing webpage on submit
        if (!this.state.description || !this.state.amount) {
            // set Error state: Please provide description and amount
            this.setState(() => ({ error: 'Please provide description and amount' }));
        } else {
            //Clear the error
            this.setState(() => ({ error: undefined }));
            // To make ExpenseForm reusable (for Add and Edit) we have to use props with function to add/edit
            // use function from parent container (props) to edit (send data to parent container)
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) *100, // parse string to a float decimals and convert from dollars to cents (* 100)
                createdAt: this.state.createdAt.valueOf(), // takes Unix Timestamp im milliseconds from moment() object
                note: this.state.note
            });
            console.log('Submited');
        }
    };
    render() {
        //console.log('PROPS from ExpenseForm.js:', this.props);
        return (
            <div>
                {/* ERROR */}
                {this.state.error && (<p>{this.state.error}</p>)}
                <form onSubmit={this.onSubmit} action="">
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    {/* Validation in HTML 5: */}
                    <input
                        type="number"
                        placeholder="Amount (HTML5 + JS Validation)"
                        min="0"
                        step=".01"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    {/* JS Validation ONLY */}
                    <input
                        type="text"
                        placeholder="Amount (JS validation ONLY)"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    {/* https://github.com/airbnb/react-dates#singledatepicker */}
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        id="DatePicker"
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    /> {/* first 5 settings of SingleDatePicker are obligatory
                    isOutsideRange gives us many advanced settings.
                    Simple function returning false let us choose past dates (by default only future */}
                    <textarea
                        name=""
                        id=""
                        cols="20"
                        rows="3"
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    ></textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        );
    }
}