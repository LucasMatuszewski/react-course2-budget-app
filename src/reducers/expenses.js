/***********
 * REDUCERS
 */

// Expanses Reducer
const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => { //default state is empty array
    switch (action.type) {
        case 'ADD_EXPENSE':
            // state.push(action.expense); // We can't use PUSH = it changes array. Reducer wold not be a PURE Function
            // state.concat(action.expense); // CONCAT() don't change array, it returns new one.
            // BUT WE CAN USE ES6 SPREAD OPERATOR:
            return [ //with spread operator we have a control where to add new item (it could be on beginning)
                ...state,
                action.expense
            ];
        case 'REMOVE_EXPENSE':
            // console.log(action.id);
            // console.log(state);
            // function removeId(expense) {
            //     console.log(expense.id);
            //     return expense.id != action.id
            // }
            // return state.filter((expense) => expense.id != action.id );
            return state.filter(({ id }) => id != action.id); // destructuring id from expense object
        // filter doesn't change array, returns new one
        case 'EDIT_EXPENSE':
            /* //my idea to do it. Works, but .map() is better :)
            console.log('state:', state.filter(({ id }) => id == action.id)[0]);
            return {
                ...state.filter(({id}) => id == action.id )[0],
                amount: action.updates
            }; */
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense, //spread previous values
                        ...action.updates //spread with new values, to edit / overwrite old values
                    }
                } else {
                    return expense;
                }
            });
        // INSTRUCTOR DIDN'T WANT US TO ADD THIS IN EXPENSES REDUCER... ONLY TO FILTERS REDUCER
        /* case 'SET_TEXT_FILTER':
            return state.filter(({ id, description, note, amount }) => {
                return id == action.text || description == action.text || note == action.text || amount == action.text;
                // return description == action.text;
            }); */
        default:
            return state;
    }
};

// export default expensesReducer; // we don't have to name this reducer if we export it by default