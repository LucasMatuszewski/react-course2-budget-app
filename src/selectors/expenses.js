import moment from 'moment';
/******
 * SELECTOR
 * GET FILTERED/Visable EXPENSES & Sort them
 */
export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        /* OLD VERSION WITHOUT MOMENT():
        const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
        const endDateMatch = typeof startDate !== 'number' || expense.createdAt <= endDate; */
        // QUERY with MOMENT() https://momentjs.com/docs/#/query/is-same-or-before/
        // IMPORTANT!!! There is some error/bug because of Redux DevTools on data changes:
        // startDate.isSameOrBefore is not a function
        // After deactivation of Redux DevTools it works fine.
        // similar isue: https://github.com/zalmoxisus/redux-devtools-extension/issues/315
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true; // startDate is already a moment() object, we don't have to convert it
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        // const textMatch = typeof text !== 'string' || expense.description.toLowerCase().indexOf(text.toLowerCase()) !== -1 || expense.note.toLowerCase().indexOf(text.toLowerCase()) !== -1;
        //in ES7 / ES2016 we have .includes(), new method better from .indexOf() in many cases:
        const textMatch = typeof text !== 'string' || expense.description.toLowerCase().includes(text.toLowerCase()) || expense.note.toLowerCase().includes(text.toLowerCase());
        // .includes() search also for NaN and undefined (indexOf don't)

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1;
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1;
        }
    });
};

// TIMESTAMPS (in milliseconds, 1s = 1000 ms )
// start 0 = 01.01.1970 (unix epoch)
// 1000 = 1000 milliseconds = 1s after 01.01.1970
// -60000 = 60s = 1min before 01.01.1970