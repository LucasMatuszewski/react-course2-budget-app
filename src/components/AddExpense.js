import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

// 1. THIS STATELESS COMPONENT HAVE INLINE FUNCTIONS. TO AVOID IT WE WILL REFACTOR IT TO CLASS COMPONENT:
// const AddExpense = (props) => (
//     <div>
//         <h1>Add Expense</h1>
//         <ExpenseForm
//             onSubmit={(expense) => {
//                 {/* Dispatch(imported action) can't be used by JEST TEST: */}
//                 {/* props.dispatch(addExpense(expense)); */}
//                 {/* WE HAVE TO MAP DISPATCH TO PROPS, and use here mapped function: */}
//                 props.onSubmit(expense);
//                 props.history.push('/'); // redirect to main page (dashboard) using browser history
//                 {/* console.log(expense); */}
//             }}
//         />
//     </div>
// );

// 2. TO AVOID INLINE FUNCTIONS WE REFACTORED STATELESS FUNCTION COMPONENT TO THIS:
export class AddExpense extends React.Component { // named export of unconnected component for JEST TESTS
    onSubmit = (expense) => {
        this.props.addExpense(expense); // action generator
        this.props.history.push('/'); // redirect to main page (dashboard) using browser history
    }
    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                />
            </div>
        )
    }
};

// Normal Component without Redux gets props only from parent Component and send props to children.
// With mapStateToProps we can connect current state of a component to Redux Store
// With mapDispatchToProps we can connect dispatching ACTIONS to Redux Store ???
// Now, our component dispatch actions and use state only by Redux Store.
// It gives us separation of concerns: Components are only to Display. Fetching data / actions are for Redux.
// We should also split components to:
// * Container (fetch data from Redux Store and send it to Presentational Child Component)
// * Presentational (display data from Container Parent Component)
// https://redux.js.org/basics/usage-with-react#presentational-and-container-components


// We don't need mapStateToProps here - we don't take data from props (e.g. existing expense)

// BUT for Jest Test we need mapDispatchToProps because we want to test:
// props.dispatch(addExpense(expense));
// But Jest don't have access to dispatch. With mapping, we can use props.onSubmit(expense) in Component.
// So lets map dispatch to our props:
/* const mapDispatchToProps = (dispatch) => {
    return {
        onSubmit: (expense) => dispatch(addExpense(expense))
    }; // NOW WE CAN USE: props.onSubmit(expense);
}; */

//SHORTER VERSION:
const mapDispatchToProps = (dispatch) => ({
    addExpense: (expense) => dispatch(addExpense(expense))
}); // NOW WE CAN USE: props.onSubmit(expense) and TEST it with JEST SPY.

// MORE ABOUT CONNECT AND MAPPING:
// https://github.com/reduxjs/react-redux/blob/master/docs/api.md#connectmapstatetoprops-mapdispatchtoprops-mergeprops-options
// https://stackoverflow.com/questions/39419237/what-is-mapdispatchtoprops

export default connect(undefined, mapDispatchToProps)(AddExpense);
// export default connect(STATE = UNDEFINED, DISPATCH)(COMPONENT TO CONNECT WITH REDUX STORE);