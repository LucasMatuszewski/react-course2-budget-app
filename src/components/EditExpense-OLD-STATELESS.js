// OLD VERSION WITH STATELESS FUNCTION COMPONENT, without mapDispatchToProps, and named exporting to JEST test
// REFACTOR MADE THE SAME WAY LIKE FOR AddExpense.js Component.

import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';
// import { removeExpense } from '../actions/expenses'; // we can import two in one line

// We have access to Props from <Route /> in AppRouter.js (like with normal React Component)

/* 
// We can use this short version, if we don't use any JS in Component (only wants to render JSX)
const EditExpense = (props) => (
    <div>
        Edit Expense
    </div>
); 

// If we want to use some JS, we have to use normal version with return(): */
const EditExpense = (props) => {
    /* // MY VERSION - filter() returns Array after filter. find() returns first ELEMENT that match!
    // https://developer.mozilla.org/pl/docs/Web/JavaScript/Referencje/Obiekty/Array/find
    // https://developer.mozilla.org/pl/docs/Web/JavaScript/Referencje/Obiekty/Array/filter
    const expense = props.expenses.filter(({ id }) => id == props.match.params.id)[0];
    console.log('PROPS:', props);
    console.log('Expense:', expense); */

    //console.log('PROPS from EditExpense.js:', props);
    return(
        <div>
            <h1>Edit Expense: {props.expense.description}</h1>
            <p>You are editing expense id: {props.expense.id}</p>
            <ExpenseForm
                expense={props.expense}
                onSubmit={ (expense) => {
                    {/* MY VERSION:
                    // props.dispatch(editExpense(props.match.params.id, expense)); */}
                    props.dispatch(editExpense(props.expense.id, expense));
                    props.history.push('/');
                }}
            />
            {/* REFACTOR: moved "remove" button from ExpenseListItem.js */}
            <button onClick={() => {
                props.dispatch(removeExpense({id: props.expense.id})); {/* ES6 Object definition shorthand for "id: id" */}
                props.history.push('/');
            }}>
                Remove
            </button>
        </div>
    );
};
/* 
In props we have access to info from React-Router:
* Browser History of URLs (with goBack, goForward), 
* Current location, with pathname
    * Search value (if we use ?query=xxx&sort=xxx in URL or ?anything=xxx)
    * Hash value from URL (if we use edukey.us#header)
* Path : URL Match - isExact: true/false
*/

// const mapStateToProps = (state) => { // MY VERSION
const mapStateToProps = (state, props) => { // UDEMY VERSION
    return {
        // expenses: state.expenses // MY VERSION
        /* expense: state.expenses.find((expense) => { // UDEMY VERSION. I used Filter inside this component
            return expense.id === props.match.params.id;
        }) */
        expense: state.expenses.find((expense) => expense.id === props.match.params.id) //SHORTER VERSION from UDEMY
    };
};

export default connect(mapStateToProps)(EditExpense);