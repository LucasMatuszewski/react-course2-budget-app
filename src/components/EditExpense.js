// NEW VERSION WITH STATEFULL CLASS COMPONENT, mapDispatchToProps and named exporting to JEST test
// REFACTOR MADE THE SAME WAY LIKE FOR AddExpense.js Component.
// OLD VERSION WITH MORE COMMENTS INSIDE: EditExpense-OLD-STATELESS.js

import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';
// import { removeExpense } from '../actions/expenses'; // we can import two in one line



export class EditExpense extends React.Component {
    onSubmit = (expense) => {
        // this.props.editExpense(expense); // MY VERSION (works fine...)
        this.props.editExpense(this.props.expense.id, expense); // UDEMY VERSION
        this.props.history.push('/');
    }
    onRemove = () => {
        // this.props.removeExpense(); // MY VERSION (works fine...)
        this.props.removeExpense({ id: this.props.expense.id });
        this.props.history.push('/');
    }
    render() {
        return(
            <div>
                <h1>Edit Expense: {this.props.expense.description}</h1>
                <p>You are editing expense id: {this.props.expense.id}</p>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={ this.onSubmit }
                />
                <button onClick={ this.onRemove }>
                    Remove
                </button>
            </div>
        );
    }
};
/* 
In props we have access to info from React-Router:
* Browser History of URLs (with goBack, goForward), 
* Current location, with pathname
    * Search value (if we use ?query=xxx&sort=xxx in URL or ?anything=xxx)
    * Hash value from URL (if we use edukey.us#header)
* Path : URL Match - isExact: true/false
*/

// const mapStateToProps = (state) => { // MY VERSION
const mapStateToProps = (state, props) => { // UDEMY VERSION
    return {
        // expenses: state.expenses // MY VERSION
        /* expense: state.expenses.find((expense) => { // UDEMY VERSION. I used Filter inside this component
            return expense.id === props.match.params.id;
        }) */
        expense: state.expenses.find((expense) => expense.id === props.match.params.id) //SHORTER VERSION from UDEMY
    };
};
/* 
// MY VERSION (worked fine, but why???)
const mapDispatchToProps = (dispatch) => ({
    editExpense: (expense) => dispatch(editExpense(props.expense.id, expense)),
    removeExpense: () => dispatch(removeExpense({id: props.expense.id}))
}); // NOW WE CAN USE: props.onSubmit(expense) and TEST it with JEST SPY.
*/

//UDEMY VERSION:
const mapDispatchToProps = (dispatch, props) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: (data) => dispatch(removeExpense(data))
}); // NOW WE CAN USE: props.onSubmit(expense) and TEST it with JEST SPY.


export default connect(mapStateToProps, mapDispatchToProps)(EditExpense);