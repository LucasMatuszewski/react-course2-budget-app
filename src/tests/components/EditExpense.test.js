import React from 'react';
import { shallow } from 'enzyme';
import { EditExpense } from '../../components/EditExpense';
import expenses from '../fixtures/expenses';

let expense, editExpense, history, removeExpense, wrapper;

beforeEach(() => { // set this 3 variables before each test case:
    expense = expenses[1]; // ON UDEMY: he use this directly as component props, and didn't let expense.
    editExpense = jest.fn(); //Spy to mock editExpense action from EditExpense component
    removeExpense = jest.fn(); //Spy to mock removeExpense action from EditExpense component
    history = { push: jest.fn() };
    wrapper = shallow(
        <EditExpense
            expense={expense}
            editExpense={editExpense}
            removeExpense={removeExpense}
            history={history}
        />
    );
});

test('should render EditExpense correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit to Edit Expense', () => {
    // 1. Sellect ExpenseForm, 2. Access prop=onSubmit, 3. Call callback function of onSubmit with (this data)
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]); // We send only one argument to this function, in EditExpense.js line 14: onSubmit = (expense) => {
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id, expenses[1]); // But this function use two arguments, in EditExpense.js line 69: dispatch(editExpense(id, expense)),
});

test('should handle onClick to Remove Expense', () => {
    wrapper.find('button').prop('onClick')(); // We don't send any arguments to this function
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(removeExpense).toHaveBeenLastCalledWith({ id: expenses[1].id }); // But this function us one argument with ID in an object { id: ID }
});