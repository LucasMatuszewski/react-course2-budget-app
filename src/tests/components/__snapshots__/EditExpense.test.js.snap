// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpense correctly 1`] = `
<div>
  <h1>
    Edit Expense: 
    Rent
  </h1>
  <p>
    You are editing expense id: 
    2
  </p>
  <ExpenseForm
    expense={
      Object {
        "amount": 1212341,
        "createdAt": -345600000,
        "description": "Rent",
        "id": "2",
        "note": "test",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
